src/lib/ngsi/EntityId.cpp

```diff
@@ -158,7 +158,7 @@ std::string EntityId::check(RequestType requestType)
   if (isTrue(isPattern))
   {
     regex_t re;
-    if (regcomp(&re, id.c_str(), REG_EXTENDED) != 0)
+    if ((regcomp(&re, id.c_str(), REG_EXTENDED) != 0) || ((id.find('\0')) != (std::string::npos)))
```

Comment:
The C++ check here is maybe too cryptic an potencially ineficient in a field that is going to be checked very often (a large set of operations in the API use entities id). Mabye @kzangeli have some C-based alternative that could be evaluated.

Reply:
In addition, maybe the check for `\0` should come first in the `||` expression, to avoid calling regcomp() if the check fails.

Reply:
@fgalan I Think find method is using the simplest possible way of finding a substring in a given string and we don't need any separate algorithm for doing the same.
For Optimizing the performance here either we can use **MACRO** instead of a function call, which will avoid stack operations of a function call in program. We will implement a normal loop in that macro which will traverse through the string and returns TRUE if substring is found else FALSE.
Or the other approach could be, We can move the same check at some other place instead of EntityID::check which will reduces the operation on entityId.
Which'll you prefer among both ?

Reply:
The cryptic part is that it is a search of ZERO in a string ... Strings can't contain zeroes.
What this really does, if I have understood correctly is to find the string `\u0000`, and in a slow way using std::string.

Would this do the trick?   `strstr(id.c_str(), "\u0000") != NULL`.

I have a few questions ...  The example functest is with `"id": "\u0000"`, i.e., an "empty" id.
Would  `"id": "E1\u0000"`also be an error?  
Both the "find implementation" and my strstr will flag an error for this.
Perhaps we should compare only the beginning of the string with `\u0000`?

One more doubt: do we support this kind of UTF chars? (`\uxxxx`)

Reply:
@kzangeli 
\>Would "id": "E1\u0000"also be an error?
Yes, it's an error as well.
\>One more doubt: do we support this kind of UTF chars? (\uxxxx)
To the best of my knowledge, this is valid in JSON. So it needs to be handled in BSON which is used by mongodb. 
\>Would this do the trick? strstr(id.c_str(), "\u0000") != NULL.
No, This won't work. I didn't got into the details yet but after making these changes program is still crashing.

Reply:
BY slightly changing your suggested command, it can be resolved.
```strstr(id.c_str(), "\0") == NULL```
This statement is working fine.